src/secur3dit/ui/UifxmlController.java

```diff
@@ -0,0 +1,374 @@
+        blurVal.setVisible(true);
+        pixelateVal.setVisible(true);
+        sharpenVal.setVisible(true);
+        rotateVal.setVisible(true);
+    }
+
+    @FXML
+    void applyWaterMark(ActionEvent e) throws IOException {
+        if (!checkImageInput()) {
+            return;
+        }
+        watermarkText = text.getText();
+        if (watermarkText == "") {
+            Alert a = new Alert(AlertType.ERROR);
+            a.setTitle("SECUR3DIT");
+            a.setHeaderText("Text is empty");
+            a.show();
+            return;
+        }
+        this.output = Filters.addWatermark(this.output, watermarkText);
+        Alert a = new Alert(AlertType.INFORMATION);
+        ImageIO.write(output, "png", new File("src/secur3dit/ui/images/output.png"));
+        a.setTitle("Secur3dit");
+        a.setHeaderText("Watermark added successfully");
+        a.show();
+        updateImage();
+
+    }
+
+    @FXML
+    void applyGrayscale(ActionEvent e) throws IOException {
+        if (!checkImageInput()) {
+            return;
+        }
+        this.output = Filters.grayscale(this.output);
+        ImageIO.write(output, "png", new File("src/secur3dit/ui/images/output.png"));
+        updateImage();
+        displayMessage();
+    }
+
+    @FXML
+    void applySepia(ActionEvent e) throws IOException {
+        if (!checkImageInput()) {
+            return;
+        }
+
+        this.output = Filters.sepia(this.output);
+        ImageIO.write(output, "png", new File("src/secur3dit/ui/images/output.png"));
+        updateImage();
+        displayMessage();
+    }
+
+    @FXML
+    void applyNegative(ActionEvent e) throws IOException {
+        if (!checkImageInput()) {
+            return;
+        }
+
+        this.output = Filters.negative(this.output);
+        ImageIO.write(output, "png", new File("src/secur3dit/ui/images/output.png"));
+        updateImage();
+        displayMessage();
+    }
+
+    @FXML
+    void applyDetectEdges(ActionEvent e) throws IOException {
+        if (!checkImageInput()) {
+            return;
+        }
+        this.output = Filters.detectEdges(this.output);
+        ImageIO.write(output, "png", new File("src/secur3dit/ui/images/output.png"));
+        displayMessage();
+        updateImage();
+    }
+
+    @FXML
+    void verticalMirrorImage(ActionEvent e) throws IOException {
+        if (!checkImageInput()) {
+            return;
+        }
+
+        this.output = Filters.mirror(this.output, true);
+        ImageIO.write(output, "png", new File("src/secur3dit/ui/images/output.png"));
+        updateImage();
+        Alert a = new Alert(AlertType.INFORMATION);
+        a.setTitle("Secur3dit");
+        a.setHeaderText("Filter applied.");
+        a.show();
+    }
+
+    @FXML
+    void horizontalMirrorImage(ActionEvent e) throws IOException {
+        if (!checkImageInput()) {
+            return;
+        }
+
+        this.output = Filters.mirror(this.output, false);
```

Comment:
Almost every method of the filters module or the crypto module throws exceptions that need to be caught by these controllers. We need try-catch blocks wherever these methods are being called.

Reply:
> Almost every method of the filters module or the crypto module throws exceptions that need to be caught by these controllers. We need try-catch blocks wherever these methods are being called.

Crypto module errors might be handled individually, but its better if for now all errors lead to a termination with an error.